src/base/bittorrent/torrentcategory.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2017  Vladimir Golovnev <glassez@yandex.ru>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <QList>
+#include <QObject>
+
+namespace BitTorrent
+{
+    class Session;
+    class TorrentHandle;
+
+    class TorrentCategory final: public QObject
+    {
+        Q_OBJECT
+        Q_DISABLE_COPY(TorrentCategory)
+
+        friend class Session;
+
+        explicit TorrentCategory(const QString &name, Session *session);
+        ~TorrentCategory() override = default;
+
+    public:
+        static QString cleanCategoryName(const QString &name);
+
+        QString name() const;
+        QString savePath() const;
+        void setSavePath(const QString &savePath);
+
+        bool contains(const TorrentCategory *category) const;
+        bool contains(const TorrentHandle *torrent) const;
+
+    signals:
+        void savePathChanged();
+
+    private:
+        Session *m_session;
```

Comment:
Why do you want to keep a pointer to a singleton?

Reply:
We already discussed such case.
I believe that the Singleton is just a use of the class, and it's wrong to refer to this class this way neither from itself, nor from its component classes.

Reply:
> I believe that the Singleton is just a use of the class…

 Does it mean that what you do with `m_session` in this class is not a "use"?

And, secondly, you have the session instance as QObject::parent(), and there is no need to duplicate the data.

Reply:
>And, secondly, you have the session instance as QObject::parent(), and there is no need to duplicate the data.

But can we trust `QObject::parent()` in case **QObject** has public `QObject::setParent(QObject*)`?
It is clear that all this is purely theoretical, because TorrentCategory is not part of some library, and hardly some of qBittorrent own devs will "guess" to do `category->setParent(someWrongObject)`, but still?..

Reply:
>Does it mean that what you do with m_session in this class is not a "use"?

Last time you did not understand me too...

Reply:
> But can we trust QObject::parent() 

`QObject` itself is not a perfect example of the class design. 

> Last time you did not understand me too...

Could you point me, please, to that previous discussion? I can't recall it.

Reply:
>`QObject` itself is not a perfect example of the class design.

Let's not discuss such things. Let the developers of Qt and the developers of other frameworks argue about it. We should take it as a given (in this project, of course). BTW, if you use `QObject::parent` property as intended, it will not cause any problems.

Reply:
>Could you point me, please, to that previous discussion? I can't recall it.

I do not remember exactly. I think it was in one of your PRs...